src/wasm/extract.cpp

```diff
@@ -374,6 +397,241 @@ std::string extractor::dump_dirs_info(json& main_dir_obj,
+}
+
+static std::vector<processed_file>
+get_renamed_collisions(const std::string& path,
+                       const std::vector<const processed_file*>& colliding_files,
+                       const processed_file* base_file, const std::string& default_language) {
+  std::vector<processed_file> renamed_collisions;
+
+  static const setup::file_entry::flags ARCH_FLAGS =
+      setup::file_entry::Bits32 | setup::file_entry::Bits64;
+
+  // We only want to append postfixes if these elements differ
+  bool common_component = true;
+  bool common_language = true;
+  bool common_arch = true;
+  for (const auto file : colliding_files) {
+    common_component =
+        common_component && file->entry().components == colliding_files[0]->entry().components;
+    common_language =
+        common_language && file->entry().languages == colliding_files[0]->entry().languages;
+    common_arch = common_arch && (file->entry().options & ARCH_FLAGS) ==
+                                     (colliding_files[0]->entry().options & ARCH_FLAGS);
+  }
+
+  for (const auto file : colliding_files) {
+    if (file == base_file) {
+      renamed_collisions.push_back(*file);
+    } else {
+      std::ostringstream oss;
+
+      bool require_number_suffix = true;
+      if (!common_component && !file->entry().components.empty()) {
+        if (setup::is_simple_expression(file->entry().components)) {
+          require_number_suffix = false;
+          oss << '#' << file->entry().components;
+        }
+      }
+      if (!common_language && !file->entry().languages.empty()) {
+        if (setup::is_simple_expression(file->entry().languages)) {
+          require_number_suffix = false;
+          if (file->entry().languages != default_language) {
+            oss << '@' << file->entry().languages;
+          }
+        }
+      }
+      if (!common_arch && (file->entry().options & ARCH_FLAGS) == setup::file_entry::Bits32) {
+        require_number_suffix = false;
+        oss << "@32bit";
+      } else if (!common_arch &&
+                 (file->entry().options & ARCH_FLAGS) == setup::file_entry::Bits64) {
+        require_number_suffix = false;
+        oss << "@64bit";
+      }
+
+      // If no postfixes were added already or the names still aren't unique,
+      // we want to differentiate them with indices
+      size_t i = 0;
+      std::string suffix = oss.str();
+      if (require_number_suffix) {
+        oss << '$' << i++;
+      }
+      while (true) {
+        if (std::none_of(renamed_collisions.cbegin(), renamed_collisions.cend(),
+                         [&](const processed_file& x) {
+                           std::string lowercase_path;
+                           std::transform(x.path().cbegin(), x.path().cend(),
+                                          std::back_inserter(lowercase_path),
+                                          [](unsigned char c) { return std::tolower(c); });
+                           return lowercase_path == path + oss.str();
```

Comment:
This is fine if it's ensured that path+oss.str() will always be lowercase. In other case, maybe both sides of the comparison should be explicitly converted to lowercase?

Reply:
It's fine. `path` is converted to lowercase in previous steps.